packages/propeller/src/StreamsSink.mts

```diff
@@ -40,59 +106,52 @@ export class StreamsSink implements Sink {
     Object.assign(this.tracingAttrs, attrs)
   }
 
-  start(signal: AbortSignal) {
+  private handleStreamCompletion(stream: Stream) {
+    for (const [batch, streams] of this.batchStreams) {
+      streams.delete(stream)
+      if (streams.size === 0) {
+        batch.onComplete()
+        this.batchStreams.delete(batch)
```

Comment:
should always be head?
if not, you might be notifying in wrong order

Reply:
Streams are processed in-order, I guess it’s possible if you’re at the tail that your processing of the last stream of batch n-1 takes longer than processing the single event in batch n, which would trigger the complete out of order. I’ll add some defensiveness